### ISSUE TITLE
New cube.grouped_by method?

### ISSUE THREAD

**Issue by rcomer** [SciTools/iris#3218]
This has come up a few times on the [Google Group](https://groups.google.com/forum/#!topic/scitools-iris/y8imBhc3VKU) lately, and again today in our Met Office Yammer.

Is it worth having a new method on the cube that can create slices of length more than 1 along a given dimension?  I.e. it would use the same `_Groupby` function that `aggregated_by` uses, but to generate a new cubelist of cubes (without any aggregation).

See the linked Google Group thread for example usage.

**Comment by duncanwp** [comment]
Could we make this a voteable feature? I found myself wanting this again today!

**Comment by rcomer** [comment]
Hi @SciTools/iris-devs, this came up on MO Yammer again today.  Does anyone have an opinion on this?  Obviously I think it's a good idea, but I'm aware of a desire to keep the cube API to a minimum (#3429).  Possibly it could be an option within the existing `slices_over` method, then then at least the _public_ API won't have grown.

If we got a "yes in principle" then a motivated user might be encouraged to take on the implementation.

**Comment by bjlittle** [comment]
@rcomer Could you give a summary here of the use case for this, with the benefits for the user and perhaps supplemented with one or more examples?

**Comment by rcomer** [comment]
Hopefully @duncanwp can describe use cases as he has wanted this at least twice.  My own was a long time ago and would now be served by `aggregated_by`.

**Comment by thomascrocker** [comment]
Hi there, I was the one that raised this recently here https://web.yammer.com/main/threads/eyJfdHlwZSI6IlRocmVhZCIsImlkIjoiOTM0MjU1NTY4NDk0NTkyIn0
My use case is calculating various climate indices on a monthly basis:

The daily data for each month has to be processed individually to produce a single value for each month in every year of my dataset. So for example a 30 year daily dataset, then becomes a 30 year monthly dataset. I don't see how to do this other than with multiple extract operations in a loop.

**Comment by zklaus** [comment]
@thomascrocker, doesn't [`cube.aggregated_by`](https://scitools.org.uk/iris/docs/latest/iris/iris/cube.html#iris.cube.Cube.aggregated_by) work for you in this case? Incidentally, I use it for the exact same use-case.

**Comment by thomascrocker** [comment]
> @thomascrocker, doesn't [`cube.aggregated_by`](https://scitools.org.uk/iris/docs/latest/iris/iris/cube.html#iris.cube.Cube.aggregated_by) work for you in this case? Incidentally, I use it for the exact same use-case.

Apologies. So it does... I got a bit muddled in my thinking....

Incidentally.. due to the size of my data aggregated by doesn't work, and I need to split it up into chunks to process in parallel. So I'm still using multiple extract operations to fire each month and year off to a separate process, but aggregated_by would work in theory if I didn't have so many grid points..

**Comment by rcomer** [comment]
@thomascrocker, which Iris version are you using?  Thanks to @zklaus, `aggregated_by` has supported lazy aggregation since Iris 2.3.  Would that help?

**Comment by thomascrocker** [comment]
> @thomascrocker, which Iris version are you using? Thanks to @zklaus, `aggregated_by` has supported lazy aggregation since Iris 2.3. Would that help?

2.4...

I'm not sure it would since it's a custom aggregator that I want to apply to aggregated_by, not one of the standard iris.analysis ones.

**Comment by thomascrocker** [comment]
Also my experience of working with iris, is that if an operator supports lazy evaluation, you benefit from your calculation being broken into chunks to avoid running out of memory, but those chunks are still calculated in serial rather than parallel.

I've always found if I want to make use of multiple CPUs to calculate in parallel I always need to explicitly invoke the parallelism manually using dask or the multiprocessor library.

**Comment by zklaus** [comment]
@thomascrocker, I have to admit that I generally take care of my dask setup explicitly in my programs, usually with at least a kind of
```python
from dask.distributed import Client
client = Client()
```

It would be interesting to understand the problem that you describe in more detail and see if we can improve the default behaviour and/or the documentation to address it.

However, I think this is out-of-scope for *this* issue, so if you are willing to help with that, please open a new ticket and tag me to follow up there.

**Comment by duncanwp** [comment]
The main reason I wanted this method was that you don't always want to aggregate over a 'grouped' coordinate, you might want to e.g. iterate over it. This is quite a common pattern in Pandas (see https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#iterating-through-groups). 

A concrete example was in wanting to save daily cubes of three-hourly data to NetCDF, I had tried:
```
iris.coord_categorisation.add_day_of_month(cube, 'time')
for day in cube.slices_over('day_of_month'):
    iris.save(cube, filename)
```
but slices_over always returns slices of length 1. 

This is doable using the existing Iris `Groupby` machinery, but its not pretty:
```
iris.coord_categorisation.add_day_of_month(c, 'time')
# Create the aggregation group-by instance.
groupby = iris.analysis._Groupby(c.coords('day_of_month'))
dimension_to_groupby = c.coord_dims(c.coord('day_of_month'))[0]
cube_slice = [slice(None, None)] * len(c.shape)
for groupby_slice in groupby.group():
    cube_slice[dimension_to_groupby] = groupby_slice
    day = c[tuple(cube_slice)]
...

Making the groupby API public would allow much more flexibilitiy in how it's used, as hopefully this simple example shows.

**Comment by zklaus** [comment]
Making the groupby api public in some way sounds sensible to me.

Note that, depending on the situation, other work arounds might be available and possibly even better performing. In the case of the three-hourly data, I would take a look at `cube.core_data().reshape((N//8, 8, -1, ...)` and manual building of appropriate time coordinates before considering any sort of looping construct.

In other cases a custom `Aggregator` with different shape maybe appropriate.

**Comment by trexfeathers** [comment]
Since this has the 'Peloton' label, it ought to have someone assigned to it. Is anyone comfortable taking this on?

If not I will move it back to 'To do' on the [Project board](https://github.com/SciTools/iris/projects/29).

**Comment by trexfeathers** [comment]
We have discussed this as a team and have regrettably decided to close the issue with no further action.

This addition would definitely be valuable in certain cases, but it has several things working against it:

* A solution to a fairly specific problem.
* Those who do need it can already achieve it, albeit through a difficult method involving some private API.
   This isn't ideal, but nor is it as bad as having no solution at all.
* Converting to public API adds to Iris' continuing maintenance overhead.
  Public API requires carefully written and updated docstrings, CI testing, and generally needs to coded more robustly, with fewer assumptions, than private API.

So in ideal circumstances we would implement this, but given the constraints on resource it just isn't possible at the moment 😔.
